code/modules/mob/living/carbon/human/species_types/jellypeople.dm

```diff
@@ -479,88 +479,275 @@
+ *
+ * Arguments
+ * * source_mob: The mob that died.
+ */
+/datum/species/oozeling/stargazer/proc/link_death(mob/living/source_mob)
+	SIGNAL_HANDLER
+	if(!source_mob.mind || !(source_mob.mind in linked_minds))
+		return
+	var/datum/weakref/mind_ref = GLOB.slime_linked_with[source_mob.mind]
+	var/datum/species/oozeling/stargazer/stargazer = mind_ref?.resolve()
+	var/mob/living/carbon/human/link_owner = stargazer?.slimelink_owner?.resolve()
+	if(!link_owner || !istype(link_owner))
+		return
+	to_chat("<span class='danger'>As you die, you feel your link to [link_owner.real_name] fizzle out!</span>")
+	unlink_mind(source_mob.mind)
+
+/**
+ * Links the mind of another mob to this stargazer's slime link.
+ *
+ * Arguments
+ * * mind: The mind to link to the slime link.
+ * Returns TRUE if the mind was successfully linked, FALSE otherwise.
+ */
+/datum/species/oozeling/stargazer/proc/link_mind(datum/mind/mind)
+	if(!validate_mind(mind, initial_connection = TRUE))
 		return FALSE
-	if(M in linked_mobs)
+	var/mob/living/carbon/human/owner = slimelink_owner.resolve()
+	if(mind in linked_minds)
 		return FALSE
 	if(!owner)
 		return FALSE
-	linked_mobs.Add(M)
-	to_chat(M, "<span class='notice'>You are now connected to [owner.real_name]'s Slime Link.</span>")
+	linked_minds.Add(mind)
 	var/datum/action/innate/linked_speech/action = new(src)
 	linked_actions.Add(action)
-	action.Grant(M)
+	action.Grant(mind.current)
+	to_chat(mind.current, "<span class='notice'>You are now connected to [owner.real_name]'s Slime Link.</span>")
+	GLOB.slime_linked_with[mind] = WEAKREF(src)
+	register_mob_signals(mind.current, death = TRUE, mind_transfer = TRUE)
+	to_chat(mind.current, "<span class='big notice'>You can use :[MODE_KEY_SLIMELINK] or .[MODE_KEY_SLIMELINK] to talk over the slime link!</span>")
 	return TRUE
 
-/datum/species/oozeling/stargazer/proc/unlink_mob(mob/living/M)
-	var/link_id = linked_mobs.Find(M)
-	if(!(link_id))
+/**
+ * Unlinks a mind from this stargazer's slime link.
+ *
+ * Arguments
+ * * mind: The mind to unlink from the slime link.
+ */
+/datum/species/oozeling/stargazer/proc/unlink_mind(datum/mind/mind)
+	var/link_id = linked_minds.Find(mind)
+	if(!link_id)
 		return
-	var/datum/action/innate/linked_speech/action = linked_actions[link_id]
-	action.Remove(M)
 	var/mob/living/carbon/human/owner = slimelink_owner.resolve()
-	if(owner)
-		to_chat(M, "<span class='notice'>You are no longer connected to [owner.real_name]'s Slime Link.</span>")
-	linked_mobs -= M
+	var/datum/action/innate/linked_speech/action = linked_actions[link_id]
+	unregister_mob_signals(mind)
+	if(mind.current)
+		action.Remove(mind.current)
+		to_chat(mind.current, "<span class='notice'>You are no longer connected to [owner.real_name]'s Slime Link.</span>")
 	linked_actions -= action
-	qdel(action)
+	linked_minds -= mind
+	if(GLOB.slime_linked_with[mind])
+		var/datum/weakref/mind_ref = GLOB.slime_linked_with[mind]
```

Comment:
weakref is not mind, but you named the variable as mind_ref. you made species type object as weakref.